### PR TITLE
Enable black formatting checks in code checks

### PR DESCRIPTION
**Pull request by elezar**
This PR updates the code checks as follows:

* Adds black formatting checks (and adds the relevant flake8 config)
* Moves the checks to pre-commit instead of running these in individual docker steps for better portability.

TODO:
- [x] Add docs on adding the pre-commit hooks
- [x] Add docs on running black

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -1,3 +1,5 @@
 [flake8]
-max-line-length = 120
-ignore = E501
+ignore = E203, E266, E501, W503
```
Comment (nickbp) [inline .flake8]: FWIW it looks like this file supports `# comments`. Could specify what the `ignored` and `select` values mean

See examples in http://flake8.pycqa.org/en/latest/user/configuration.html

Comment (elezar) [reply .flake8]: Thanks. Will do.

Comment (elezar) [reply .flake8]: 6360452

File: .flake8 [hunk 2]
```diff
@@ -1,3 +1,12 @@
 [flake8]
-max-line-length = 120
-ignore = E501
+# The following are ignored:
+#   E203 whitespace before ':' (https://github.com/ambv/black#line-breaks--binary-operators)
+#   E501 line too long         (https://github.com/ambv/black#line-length)
+#   W503 line break before binary operator
+ignore = E203, E501, W503
+max-line-length = 88
```
Comment (porridge) [inline .flake8]: Noooo... could we set black to make it at least 100 please? (Personally I'd even vote for keeping 120, but the rationale seems https://github.com/ambv/black#line-length kind-of convinces me...)

File: .flake8 [hunk 3]
```diff
@@ -1,3 +1,12 @@
 [flake8]
-max-line-length = 120
-ignore = E501
+# The following are ignored:
+#   E203 whitespace before ':' (https://github.com/ambv/black#line-breaks--binary-operators)
+#   E501 line too long         (https://github.com/ambv/black#line-length)
```
Comment (porridge) [inline .flake8]: What's the point of both setting the line length below *and* ignoring the warnings?

File: .flake8 [hunk 4]
```diff
@@ -1,3 +1,12 @@
 [flake8]
-max-line-length = 120
-ignore = E501
+# The following are ignored:
```
Comment (porridge) [inline .flake8]: It's not obvious right away why are these settings in `[flake8]` section but the links in comments point to `black`. Perhaps add ` to work better with "black"` before the colon.

File: .flake8 [hunk 5]
```diff
@@ -1,3 +1,12 @@
...
+#   E203 whitespace before ':' (https://github.com/ambv/black#line-breaks--binary-operators)
```
Comment (porridge) [inline .flake8]: Change the link to https://github.com/ambv/black#slices

File: .flake8 [hunk 6]
```diff
@@ -1,3 +1,12 @@
...
+#   E501 line too long         (https://github.com/ambv/black#line-length)
+#   W503 line break before binary operator
```
Comment (porridge) [inline .flake8]: Link to https://github.com/ambv/black#line-breaks--binary-operators

File: tools/ci/checks/run_black_checks.sh [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+DOCKER_TAG=${DOCKER_TAG:-latest}
+DOCKER_IMAGE=${DOCKER_IMAGE:-mesosphere/dcos-commons:${DOCKER_TAG}}
+
+if [ x"$1" == x"--format" ]; then
```
Comment (porridge) [inline tools/ci/checks/run_black_checks.sh]: Please drop the x, or drop the quotes and use ``[[ ]]` better yet.

File: .flake8 (hunk 3 above)
Comment (elezar) [reply .flake8]: Hmmmm. Black doesn't always change the line length. Some examples I've seen is long doc-strings and string parameters:
```
✗ black testing; flake8 testing | wc -l
All done! ✨ 🍰 ✨
22 files left unchanged.
     157
```

File: tools/ci/checks/run_black_checks.sh (hunk 7 above)
Comment (elezar) [reply tools/ci/checks/run_black_checks.sh]: I will drop the file.

File: .flake8 (hunk 2 above)
Comment (elezar) [reply .flake8]: Ok. I was going with the defaults for black, but 100 is a good compromise.

**Comment by elezar** [comment]
Thanks @nickbp and @porridge.

**Comment by nickbp** [comment]
FWIW unless anyone else has comments I think this is ready to go. The cassandra failure is due to a known regression in metronome and shouldn't block merging.

**Comment by elezar** [comment]
> FWIW unless anyone else has comments I think this is ready to go. The cassandra failure is due to a known regression in metronome and shouldn't block merging.

Thanks. I've just updated with Master and will merge as soon as all except Cassandra are green.

**Comment by elezar** [comment]
I'm merging this. As discussed, the Cassandra issues are known, and there are no changes in this PR that would affect any framework.

**Pull request merged**